src/samples/iio+network/iio-magnetometer-console-and-mqtt-publish.c

```diff
@@ -0,0 +1,907 @@
+    if (data[0] == 0 || data[1] == 0 || data[2] == 0)
+        return -1;
+
+    lookback_count = lookback_counts[cal_data->cal_level];
+    min_diff = min_diffs[cal_data->cal_level];
+
+    /* For the current point to be accepted, each x/y/z value must be different
+     * enough to the last several collected points
+     */
+    if (cal_data->sample_count > 0 && cal_data->sample_count < MAGN_DS_SIZE) {
+        unsigned int lookback =
+            lookback_count < cal_data->sample_count ? lookback_count : cal_data->sample_count;
+        for (index = 0; index < lookback; index++)
+            for (j = 0; j < 3; j++)
+                if (fabsf(data[j] - cal_data->sample[cal_data->sample_count - 1 - index][j]) <
+                    min_diff) {
+#ifdef DEBUG
+                    printf("CompassCalibration:point reject: [%f,%f,%f], selected_count=%d\n",
+                        data[0],
+                        data[1],
+                        data[2],
+                        cal_data->sample_count);
+#endif
+                    return 0;
+                }
+    }
+
+    if (cal_data->sample_count < MAGN_DS_SIZE) {
+        memcpy(cal_data->sample[cal_data->sample_count], data, sizeof(double) * 3);
+        cal_data->sample_count++;
+        cal_data->average[0] += data[0];
+        cal_data->average[1] += data[1];
+        cal_data->average[2] += data[2];
+#ifdef DEBUG
+        printf("CompassCalibration:point collected [%f,%f,%f], selected_count=%d\n",
+            data[0],
+            data[1],
+            data[2],
+            cal_data->sample_count);
+#endif
+    }
+    return 1;
+}
+
+static void
+compass_compute_cal(struct compass_cal *cal_data, double *x, double *y, double *z)
+{
+    double result[3][1], raw[3][1], diff[3][1];
+
+    if (!cal_data->cal_level)
+        return;
+
+    raw[0][0] = *x;
+    raw[1][0] = *y;
+    raw[2][0] = *z;
+
+    substract(3, 1, raw, cal_data->offset, diff);
+    multiply(3, 3, 1, cal_data->w_invert, diff, result);
+
+    *x = result[0][0];
+    *y = result[1][0];
+    *z = result[2][0];
+
+    scale(x, y, z);
+}
+
+static void
+calibrate_compass(struct compass_cal *cal_data, double *x, double *y, double *z)
+{
+    /* Thanks to https://github.com/01org/android-iio-sensors-hal for calibration algorithm */
+
+    /* Calibration is continuous */
+    compass_collect(cal_data, x, y, z);
+
+    compass_ready(cal_data);
+
+    if (cal_data->cal_level == 0)
+        scale(x, y, z);
+    else
+        compass_compute_cal(cal_data, x, y, z);
+}
+#endif
+
+#ifdef DENOISE_AVERAGE
+static void
+denoise_average(int sampling_frequency, struct filter_average *filter, double *x, double *y, double *z)
+{
+    /* Thanks to https://github.com/01org/android-iio-sensors-hal for denoise algorithm */
+    /*
+     * Smooth out incoming data using a moving average over a number of
+     * samples. We accumulate one second worth of samples, or max_samples,
+     * depending on which is lower.
+     */
+    double *data[3];
```

Comment:
the correct way would be you allocate it using calloc with `filter->num_fields` to avoid access invalid memory address. If you want keep it on stack, at least change it to
```
double *data[FILTER_NUM_FIED]
```